### PR TITLE
Polygon triangulation supports collinear vertices

### PR DESCRIPTION
**Pull request by lucas-flexcompute**
This is a fix for our ear clipping triangulation algorithm to work with polygons with collinear vertices.
It also adds a safeguard in the main loop to error if no progress is made, so that we don't get stuck jobs due to numerical issues.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - Users can manually specify the background medium for a structure to be used for geometry gradient calculations by supplying `Structure.background_permittivity`. This is useful when there are overlapping structures or structures embedded in other mediums.
 
+### Changed
+- Triangulation of `PolySlab` poolygons now supports pollygons with collinear vertices.
```
Comment (momchil-flex) [inline CHANGELOG.md]: Two different typos in `polygons` :)

**Review by momchil-flex — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (lucas-flexcompute) [reply CHANGELOG.md]: I'm blaming that on a different keyboard that I'm using these past few days... :/

**Comment by momchil-flex** [comment]
Awesome btw tested that this un-stucks the problematic sim! You probably did as well but I verified.

File: CHANGELOG.md (hunk 1 above)
Comment (weiliangjin2021) [reply CHANGELOG.md]: It happens to me too with my apple magic keyboard that has been with me for 10 years. Now have to switch to a new one.

File: tidy3d/components/geometry/triangulation.py [hunk 2]
```diff
@@ -41,15 +42,21 @@ def update_convexity(vertices: List[Vertex], i: int) -> None:
         Vertices of the polygon.
     i : int
         Index of the vertex to be updated.
+
+    Returns
+    -------
+    int
+        -1 if vertex was collinear, +1 if it became collinear, 0 otherwise.
```
Comment (weiliangjin2021) [inline tidy3d/components/geometry/triangulation.py]: This part is hard to understand. Maybe add more explanation here?
My understanding by reading the code below is: -1 for vertex was collinear, and no longer collinear after the update. But what about +1 and 0?

**Review by weiliangjin2021 — APPROVED** [review]

File: tidy3d/components/geometry/triangulation.py (hunk 2 above)
Comment (lucas-flexcompute) [reply tidy3d/components/geometry/triangulation.py]: This return value is just to quickly update the counter of collinear vertices, i.e., if the update decreased the number of collinear vertices (a vertex was collinear and no longer is), it returns -1; if the update increased the number of collinears (the vertex was not collinear and now it is), it returns +1; else, the update doesn't change the count and it returns 0.

Comment (momchil-flex) [reply tidy3d/components/geometry/triangulation.py]: Maybe just add this (or a short version of it) to the docstring @lucas-flexcompute ?

Comment (lucas-flexcompute) [reply tidy3d/components/geometry/triangulation.py]: Done! I think it's more clear now with the note and comment.

Comment (momchil-flex) [reply tidy3d/components/geometry/triangulation.py]: Thanks, could you squash and merge.

**Pull request merged**